codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -16,329 +16,514 @@
+            LOG.warn("Stream exception thrown for unkown stream.", cause);
+            // Write a RST_STREAM
+            super.onStreamError(ctx, cause, streamException);
+            return;
         }
 
-        @Override
-        protected void onStreamError(ChannelHandlerContext ctx, Throwable cause,
-                                     Http2Exception.StreamException http2Ex) {
-            try {
-                Http2Stream stream = connection().stream(http2Ex.streamId());
-                if (stream == null) {
-                    return;
-                }
-                ctx.fireExceptionCaught(http2Ex);
-            } finally {
-                super.onStreamError(ctx, cause, http2Ex);
-            }
+        Http2FrameStream stream = connectionStream.getProperty(streamKey);
+        if (stream == null) {
+            LOG.warn("Stream exception thrown without stream object attached.", cause);
+            // Write a RST_STREAM
+            super.onStreamError(ctx, cause, streamException);
+            return;
         }
+
+        onHttp2FrameStreamException(ctx, new Http2FrameStreamException(stream, streamException.error(), cause));
+    }
+
+    @Override
+    protected final boolean isGracefulShutdownComplete() {
+        return super.isGracefulShutdownComplete() && numBufferedStreams == 0;
     }
 
-    private static final class FrameListener extends Http2FrameAdapter {
+    private final class FrameListener extends Http2FrameAdapter {
+
+        @Override
+        public void onUnknownFrame(
+                ChannelHandlerContext ctx, byte frameType, int streamId, Http2Flags flags, ByteBuf payload) {
+            onHttp2Frame(ctx, new DefaultHttp2UnkownFrame(frameType, flags, payload.retain())
+                    .stream(requireStream(streamId)));
+        }
+
+        @Override
+        public void onSettingsRead(ChannelHandlerContext ctx, Http2Settings settings) {
+            onHttp2Frame(ctx, new DefaultHttp2SettingsFrame(settings));
+        }
+
+        @Override
+        public void onPingRead(ChannelHandlerContext ctx, ByteBuf data) {
+            onHttp2Frame(ctx, new DefaultHttp2PingFrame(data.retain(), false));
+        }
+
+        @Override
+        public void onPingAckRead(ChannelHandlerContext ctx, ByteBuf data) {
+            onHttp2Frame(ctx, new DefaultHttp2PingFrame(data.retain(), true));
+        }
+
         @Override
         public void onRstStreamRead(ChannelHandlerContext ctx, int streamId, long errorCode) {
-            Http2ResetFrame rstFrame = new DefaultHttp2ResetFrame(errorCode);
-            rstFrame.streamId(streamId);
-            ctx.fireChannelRead(rstFrame);
+            onHttp2Frame(ctx, new DefaultHttp2ResetFrame(errorCode).stream(requireStream(streamId)));
+        }
+
+        @Override
+        public void onWindowUpdateRead(ChannelHandlerContext ctx, int streamId, int windowSizeIncrement) {
+            if (streamId == 0) {
+                // Ignore connection window updates.
```

Comment:
I'd agree that it's not generally useful and so probably wasn't propagated. Can we leave it un-propagated until we actually implement it for a use-case? I'm not comfortable adding features "because we can" without an understanding of how it would be used; that tends to bite me later for new features when I am avoiding breaking applications.